internal/status/status.go

```diff
@@ -126,12 +126,12 @@ func (s *Status) Details() []interface{} {
 	}
 	details := make([]interface{}, 0, len(s.s.Details))
 	for _, any := range s.s.Details {
-		detail := &ptypes.DynamicAny{}
-		if err := ptypes.UnmarshalAny(any, detail); err != nil {
+		detail, err := any.UnmarshalNew()
+		if err != nil {
 			details = append(details, err)
 			continue
 		}
-		details = append(details, detail.Message)
+		details = append(details, detail)
```

Comment:
If I didn't miss anything, this will break all the users using the old proto library (and code generated by the old codegen), right?

The type of `detail` is changed from `v1proto.Message` to a `v2proto.Message`. Those two are different interfaces, and type assertion to `v1proto.Message` will fail after this change.

In that case, this is an API change (though it doesn't break at building time, it will break at runtime).
We can add a `DetailsV2()` method, but that means we are stuck with the v1 proto dependency.

Reply:
@menghanl Yes, you are right. This commit is a breaking change. 

However, I think this is inevitable as [github.com/golang/protobuf](github.com/golang/protobuf) has been officially marked as deprecated. Eventually, this repository will have to migrate from [github.com/golang/protobuf](github.com/golang/protobuf) to [google.golang.org/protobuf](google.golang.org/protobuf) at some point.

If you think we should keep the compatibility for v1 Message, I can convert `detail` into `v1proto.Message` by calling the [`MessageV1`](https://pkg.go.dev/github.com/golang/protobuf@v1.5.2/proto#MessageV1) function. Please let me know what do you think is best for this scenario. Thanks.

Reply:
I think in this PR, the best we can do is to add a `DetailsV2()` with the new code you have, and mark `Details()` as deprecated. In your code, you can start to use `DetailsV2()` with v2proto.

If we really want to `Details()`, we can add a `DetailsV1()`, and make `Details()` an alias of `DetailsV1()`. We need to keep all of them for a couple of releases, then we switch `Details()` to an alias of `DetailsV2()`, and eventually we can delete `DetailsV1()` and `DetailsV2()` (after a couple more releases).

I would actually be OK to keep `Details()` and `DetailsV2()` for now. The reason to get rid of `Details()` is to remove the dependency of the old proto library. But we won't be able to completely remove the dependency until we also delete v1proto from all the other packages in the grpc module (#3554)

Reply:
@menghanl I have added `DetailsV2()` and made the changes accordingly. Please review again. Thank you.

Reply:
What is the value of the V2 version?  The original version will interoperate with the new library.  Can you provide some sample code that uses `google.golang.org/protobuf` and has to interact with `Details`, and show how it is better when switching to `DetailsV2`?  I'd like to better understand what is being gained here.

> However, I think this is inevitable as github.com/golang/protobuf has been officially marked as deprecated.

I don't see how it can ever be removed or have support ended for it, however.  It would break far too many people.

> If we really want to Details(), we can add a DetailsV1(), and make Details() an alias of DetailsV1(). We need to keep all of them for a couple of releases, then we switch Details() to an alias of DetailsV2(), and eventually we can delete DetailsV1() and DetailsV2() (after a couple more releases).

None of this is labeled as experimental, so I don't see how we can do this.

The only reason to ever remove `Details` would be if the protobuf team outright deletes their repo.  In that case I think we would be okay to break users within a major release (without doing so, their code literally could not compile).

Reply:
@dfawley Here is some sample code.

```go
import (
        // other imports omitted for brevity
	protoV1 "github.com/golang/protobuf/proto"
	protoV2 "google.golang.org/protobuf/proto"
)

func main() {
	// greeter_client example code
	conn, err := grpc.Dial(address, grpc.WithInsecure(), grpc.WithBlock())
	if err != nil {
		log.Fatalf("did not connect: %v", err)
	}
	defer conn.Close()
	c := pb.NewGreeterClient(conn)
	
	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	defer cancel()

	// assume that an error is returned here, and the error can be converted to status
	r, err = c.SayHello(ctx, &pb.HelloRequest{Name: name})

	st := status.Convert(err)

	// difference between Details() and DetailsV2()
	for _, detail := range st.Details() {
		switch t := detail.(type) {
		case protoV1.Message:
			t.String()
			t.Reset()
			t.ProtoMessage()
		}
	}


	for _, detail := range st.DetailsV2() {
		switch t := detail.(type) {
		case protoV2.Message:
			t.ProtoReflect()
		}
	}
}
```

I think most of the end-users would not notice the difference unless they are doing something like the type assertion in the sample code above.

I am not a Go nor gRPC expert, hence I could not summarize any real benefits of doing this except for removing deprecated dependency and using newer protobuf API.

Reply:
I'm fine with the ptypes->anypb change, since that appears 100% identical, but I am opposed to adding a new API here if it will not provide any meaningful difference to our end users.  "Deprecated" does not mean "going away soon".

@dsnet does the protobuf team have any guidance published (or unpublished) regarding updating existing APIs that use `github.com/golang/protobuf/proto.Message`?  Actually, any kind of general guidance about the two versions would be appreciated as well.  A lot of our users have filed issues related to this, and it would be nice if we had an official policy & consistent story.  I don't want to contribute to any fragmentation, but I also don't want to complicate our API, either.

Another option, if using `github.com/golang/protobuf` is or will be a problem in the future, would be a different API that deals in raw `[]byte`s instead of proto messages.  This would isolate us from any similar changes in the future, and is what we would have done, had protobuf not already been a dependency of grpc-go.  (Note that neither grpc-java nor grpc-c have protobuf dependencies.)

Reply:
@dfawley The original version of this pull request modified `Details()` directly. However, @menghanl suggested it might be better to add `DetailsV2()` to prevent breaking change so I just followed his advice.

Reply:
Hmm, I missed that the generated message types implement both v1proto and v2proto interfaces (I was under the impression that the messages in details cannot be type asserted to the v2proto messages, so that the users are forced to use v1proto, but I was wrong).

So we don't need this change. Users using v2proto can just type assert the details to the generated concrete message types. Everything done here is transparent to the users.

Reply:
Seems what I was thinking was backwards? What's going to break is not `Details()`, but `WithDetails()`?

We cannot change the imported package from `"github.com/golang/protobuf/proto"` to `"google.golang.org/protobuf/proto"`,

because this changes the type `proto.Message` from `v1proto.Message` to `v2proto.Message`

thus it will change the function signature of `WithDetails()` (from `func(...v1proto.Message)` to `func(...v2proto.Message)`).

If a user is using very old generated code, the message types (e.g. `hellopb.Request`) will only implement `v1proto.Message` but not `v2proto.Message`. When they call `WithDetails(&hellopb.Request{})`, it will fail, because the types don't match.

Reply:
This specific API is a bit strange since it returns the message through an `interface{}` rather than a `protov1.Message` or a `protov2.Message`. Almost every concrete type that implements `protov2.Message` also implement `protov1.Message`. However, many concrete types that implement `protov1.Message` may not implement `protov2.Message` since they may be generated by `protoc-gen-gogo` or by an older version of `protoc-gen-go`.

The main difference between `ptypes` and `anypb` is how it handles cases where the concrete message type does not implement `protov2.Message`. In such cases, `anypb` wraps the message with an opaque wrapper type. The original concrete type the implements `protov1.Message` can be retrieved by calling [`protov1.MessageV1`](https://pkg.go.dev/github.com/golang/protobuf/proto#MessageV1).

Thus, the old implementation of:
```go
	detail := &ptypes.DynamicAny{}
	if err := ptypes.UnmarshalAny(any, detail); err != nil {
		details = append(details, err)
		continue
	}
	details = append(details, detail.Message)
```
would be semantically identical to:
```go
	if m, err := any.UnmarshalNew(); err != nil {
		details = append(details, err)
	} else {
		details = append(details, protov1.MessageV1(m))
	}
```
Note the presence of the `protov1.MessageV1` adaptor.

----

> [does the] protobuf team have any guidance published (or unpublished) regarding updating existing APIs that use `github.com/golang/protobuf/proto.Message`?

There's no clear answer. It comes down to whether the extra API is worth the cost for the user to do *possible* conversions. Many time conversions is not necessary.

**Case 1: `protov1.Message` appears as an argument (e.g., `func Set(protov1.Message)`)**

* In most cases, you have a concrete message on hand. In the future, most generated messages will implement `protov1.Message` and `protov2.Message`, so they can call this function without any problems (e.g., `Set(m)`).
* In some cases, you have a `protov1.Message` on hand where the type is identical, so you can call it directly (e.g., `Set(m)`).
* In some cases, you have a `protov2.Message` on hand, so you can't call it directly and must go through the `protov1.MessageV1` adaptor function (e.g., `Set(protov1.MessageV1(m))`).

**Case 2: `protov1.Message` appears as an return value (e.g., `func Get() protov1.Message`)**

* In some cases, you want to type assert to a concrete type, in which cases, there's not much of a problem since nearly every concrete message type that implements `protov2.Message` also implements `protov1.Message`.
* In some cases, you want to perform a general operation (e.g., `proto.Marshal`, `proto.Equal`, etc) in which case, people would be forced to either call `protov1.Marshal(m)` or `protov2.Marshal(protov1.MessageV2(m))`. Either way, they have a dependency on `protov1`.

**Other cases**

The above are not the only two cases, but the most common. Another case, is a function that accepts a `[]protov1.Message`, where constructing this type forces you to reference the `protov1` package (unless Go had https://github.com/golang/go/issues/12854). In such a case, it might make sense to add API that accepts `[]protov2.Message` to ease the burden on users.

Reply:
Something I'm getting out of this discussion is that we should probably add the [`MessageV1`](https://pkg.go.dev/github.com/golang/protobuf/proto#MessageV1) and [`MessageV2`](https://pkg.go.dev/github.com/golang/protobuf/proto#MessageV2) adaptor functions to `protov2` so that people won't need to depend on `protov1` if that's all they need.